File: src/ocaml-output/FStar_TypeChecker_PatternInference.ml
open Prims
let (filter_trigger :
  FStar_Syntax_Syntax.bv FStar_Util.set ->
    FStar_Syntax_Syntax.term ->
      FStar_Syntax_Syntax.term FStar_Pervasives_Native.option)
  =
  fun names1  ->
    fun t  ->
      let bvs = FStar_Syntax_Free.names t  in
      let uu____19 = FStar_Util.set_is_subset_of names1 bvs  in
      if uu____19
      then FStar_Pervasives_Native.Some t
      else FStar_Pervasives_Native.None
  
let (has_no_smt_theory_symbols :
  FStar_TypeChecker_Env.env -> FStar_Syntax_Syntax.term -> Prims.bool) =
  fun env  ->
    fun e  ->
      let uu____38 =
        let uu____39 = FStar_Syntax_Subst.compress e  in
        uu____39.FStar_Syntax_Syntax.n  in
      match uu____38 with
      | FStar_Syntax_Syntax.Tm_fvar fv when
          FStar_TypeChecker_Env.fv_has_attr env fv
            FStar_Parser_Const.smt_theory_symbol_attr_lid
          -> false
      | FStar_Syntax_Syntax.Tm_uinst
          ({ FStar_Syntax_Syntax.n = FStar_Syntax_Syntax.Tm_fvar fv;
             FStar_Syntax_Syntax.pos = uu____46;
             FStar_Syntax_Syntax.vars = uu____47;_},uu____48)
          when
          FStar_TypeChecker_Env.fv_has_attr env fv
            FStar_Parser_Const.smt_theory_symbol_attr_lid
          -> false
      | uu____54 -> true
  
let rec (find_trigger :
  FStar_TypeChecker_Env.env ->
    FStar_Syntax_Syntax.bv FStar_Util.set ->
      FStar_Syntax_Syntax.term ->
        (Prims.bool * FStar_Syntax_Syntax.term Prims.list))
  =
  fun env  ->
    fun names1  ->
      fun t  ->
        let debug1 = false  in
        if debug1
        then
          (let uu____95 = FStar_Syntax_Print.term_to_string t  in
           let uu____97 =
             let uu____99 = FStar_Syntax_Subst.compress t  in
             FStar_Syntax_Print.tag_of_term uu____99  in
           FStar_Util.print2 "find trigger for : %s (%s)\n" uu____95 uu____97)
        else ();
        (let uu____103 =
           let uu____104 = FStar_Syntax_Subst.compress t  in
           uu____104.FStar_Syntax_Syntax.n  in
         match uu____103 with
         | FStar_Syntax_Syntax.Tm_bvar x -> (false, [])
         | FStar_Syntax_Syntax.Tm_name x -> (false, [])
         | FStar_Syntax_Syntax.Tm_fvar fv -> (false, [])
         | FStar_Syntax_Syntax.Tm_uinst (e,universes) ->
             find_trigger env names1 e
         | FStar_Syntax_Syntax.Tm_constant c -> (false, [])
         | FStar_Syntax_Syntax.Tm_type u -> (true, [])
         | FStar_Syntax_Syntax.Tm_abs (xs,body,uu____147) -> (true, [])
         | FStar_Syntax_Syntax.Tm_arrow (xs,body) -> (true, [])
         | FStar_Syntax_Syntax.Tm_refine (x,phi) -> (true, [])
         | FStar_Syntax_Syntax.Tm_app (e,args) ->
             let uu____238 =
               let uu____239 = FStar_Syntax_Subst.compress e  in
               uu____239.FStar_Syntax_Syntax.n  in
             (match uu____238 with
              | FStar_Syntax_Syntax.Tm_fvar fv when
                  (FStar_Syntax_Syntax.fv_eq_lid fv
                     FStar_Parser_Const.forall_lid)
                    ||
                    (FStar_Syntax_Syntax.fv_eq_lid fv
                       FStar_Parser_Const.exists_lid)
                  -> (true, [])
              | FStar_Syntax_Syntax.Tm_uinst
                  ({ FStar_Syntax_Syntax.n = FStar_Syntax_Syntax.Tm_fvar fv;
                     FStar_Syntax_Syntax.pos = uu____255;
                     FStar_Syntax_Syntax.vars = uu____256;_},uu____257)
                  when
                  (FStar_Syntax_Syntax.fv_eq_lid fv
                     FStar_Parser_Const.forall_lid)
                    ||
                    (FStar_Syntax_Syntax.fv_eq_lid fv
                       FStar_Parser_Const.exists_lid)
                  -> (true, [])
              | uu____266 ->
                  let uu____267 = find_trigger env names1 e  in
                  (match uu____267 with
                   | (trigger_killer,uu____283) ->
                       let uu____290 =
                         FStar_List.fold_left
                           (fun uu____320  ->
                              fun uu____321  ->
                                match (uu____320, uu____321) with
                                | ((k,l),(e1,q)) ->
                                    let uu____382 =
                                      find_trigger env names1 e1  in
                                    (match uu____382 with
                                     | (b,c) ->
                                         ((k || b), (FStar_List.append l c))))
                           (trigger_killer, []) args
                          in
                       (match uu____290 with
                        | (trigger_killer1,c) ->
                            let c1 =
                              FStar_All.pipe_right c
                                (FStar_List.choose
                                   (fun t1  -> filter_trigger names1 t1))
                               in
                            let c2 =
                              match c1 with
                              | [] ->
                                  let uu____440 =
                                    (Prims.op_Negation trigger_killer1) &&
                                      (has_no_smt_theory_symbols env e)
                                     in
                                  if uu____440
                                  then FStar_List.append c1 [t]
                                  else c1
                              | uu____447 -> c1  in
                            (trigger_killer1, c2))))
         | FStar_Syntax_Syntax.Tm_match (e,branches) -> (true, [])
         | FStar_Syntax_Syntax.Tm_ascribed uu____495 -> (true, [])
         | FStar_Syntax_Syntax.Tm_let ((is_rec,source_lbs),body) ->
             (true, [])
         | FStar_Syntax_Syntax.Tm_uvar (u,uu____551) -> (false, [])
         | FStar_Syntax_Syntax.Tm_delayed uu____572 ->
             failwith "Tm_delayed is impossible after compress"
         | FStar_Syntax_Syntax.Tm_meta (e,m) -> (true, [])
         | FStar_Syntax_Syntax.Tm_lazy i -> (true, [])
         | FStar_Syntax_Syntax.Tm_quoted (tm,qi) -> (true, [])
         | FStar_Syntax_Syntax.Tm_unknown  -> (true, []))
  
let (terms_to_bvs :
  FStar_Syntax_Syntax.term Prims.list ->
    FStar_Syntax_Syntax.bv FStar_Util.set)
  =
  fun names1  ->
    match names1 with
    | [] -> failwith "Empty bound vars"
    | hd1::tl1 ->
        let uu____653 = FStar_Syntax_Free.names hd1  in
        FStar_List.fold_left
          (fun out  ->
             fun x  ->
               let uu____665 = FStar_Syntax_Free.names x  in
               FStar_Util.set_union out uu____665) uu____653 tl1
  
let (infer_pattern :
  FStar_TypeChecker_Env.env ->
    FStar_Syntax_Syntax.term Prims.list ->
      FStar_Syntax_Syntax.term -> FStar_Syntax_Syntax.args Prims.list)
  =
  fun env  ->
    fun names1  ->
      fun t  ->
        let debug1 = true  in
        if debug1
        then
          (let uu____697 = FStar_Syntax_Print.term_to_string t  in
           let uu____699 =
             let uu____701 = FStar_Syntax_Subst.compress t  in
             FStar_Syntax_Print.tag_of_term uu____701  in
           let uu____702 =
             let uu____704 =
               FStar_All.pipe_right names1
                 (FStar_List.map FStar_Syntax_Print.term_to_string)
                in
             FStar_All.pipe_right uu____704 (FStar_String.concat ", ")  in
           FStar_Util.print3 "Infer pattern for : %s (%s) with names: (%s)\n"
             uu____697 uu____699 uu____702)
        else ();
        (let bvs = terms_to_bvs names1  in
         let uu____725 = find_trigger env bvs t  in
         match uu____725 with
         | (uu____735,p) ->
             let pats =
               FStar_All.pipe_right p
                 (FStar_List.choose (fun t1  -> filter_trigger bvs t1))
                in
             (if debug1
              then
                (let uu____754 =
                   let uu____756 =
                     FStar_All.pipe_right pats
                       (FStar_List.map FStar_Syntax_Print.term_to_string)
                      in
                   FStar_All.pipe_right uu____756 (FStar_String.concat "; ")
                    in
                 FStar_Util.print1 "Found patterns: %s\n" uu____754)
              else ();
              FStar_List.fold_left
                (fun l  ->
                   fun t1  ->
                     let uu____787 =
                       let uu____792 =
                         let uu____795 = FStar_Syntax_Syntax.as_arg t1  in
                         [uu____795]  in
                       [uu____792]  in
                     FStar_List.append l uu____787) [] pats))
  
let (remove_invalid_pattern :
  FStar_Syntax_Syntax.term Prims.list ->
    FStar_Syntax_Syntax.args Prims.list ->
      FStar_Syntax_Syntax.args Prims.list)
  =
  fun names1  ->
    fun pats  ->
      match names1 with
      | [] -> pats
      | uu____827 ->
          let bvs = terms_to_bvs names1  in
          let pats1 =
            FStar_List.fold_left
              (fun l  ->
                 fun p  ->
                   let p1 =
                     FStar_List.choose
                       (fun uu____873  ->
                          match uu____873 with
                          | (t,uu____883) -> filter_trigger bvs t) p
                      in
                   FStar_List.append l p1) [] pats
             in
          FStar_List.fold_left
            (fun l  ->
               fun t  ->
                 let uu____901 =
                   let uu____906 =
                     let uu____909 = FStar_Syntax_Syntax.as_arg t  in
                     [uu____909]  in
                   [uu____906]  in
                 FStar_List.append l uu____901) [] pats1
  

File: src/typechecker/FStar.TypeChecker.PatternInference.fs
﻿(*
   Copyright 2008-2014 Nikhil Swamy and Microsoft Research

   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*)
#light "off"
module FStar.TypeChecker.PatternInference
open FStar.ST
open FStar.All

open FStar
open FStar.Util
open FStar.Syntax
open FStar.Syntax.Syntax
open FStar.TypeChecker.Env
module Const = FStar.Parser.Const
module S  = FStar.Syntax.Syntax
module SS = FStar.Syntax.Subst
module BU = FStar.Util

let filter_trigger (names:set<bv>) (t:S.term) =
    let bvs = FStar.Syntax.Free.names t in
    if BU.set_is_subset_of names bvs then Some t else None

let has_no_smt_theory_symbols env e =
    match (SS.compress e).n with
    | Tm_fvar fv
    | Tm_uinst({n=Tm_fvar fv}, _)
      when Env.fv_has_attr env fv Const.smt_theory_symbol_attr_lid -> false
    | _ -> true

let rec find_trigger env (names:set<bv>) (t:S.term) : (bool * list<S.term>) =
    (* To automatically generate patterns for (forall (x1...xn). e) or (exists (x1...xn). e), 
     F* searches inside e for a term t that contains all the variables x1...xn:
      - For simplicity, the search need not look inside other binders, 
        such as nested foralls, exists, let, or lambdas inside e.
         -- this could be improved, but it will prevent accidentally generating patterns that 
            include out-of-scope variables
      - The term t should contain only function applications, constants, and variables.  
        It should contain at least one function application (i.e. it can't just be a variable by itself).  
        It should not contain:
         -- any function with the "[@smt_theory_symbol]" attribute, like /\, ==, +, etc.
         -- "let", foralls, exists, and lambdas
      - F* should make t as small as possible while still covering x1...xn.  
        For example, it should choose f(x, y) instead of h(f(x, y)).
      - If there are no such t, automatic generation fails.
      - There may be more than one t.  In this case, F* would generate a disjunction of the 
        terms {:pattern t1 \/ ... \/ tk}. *)
    let debug = false in
    if debug then 
      BU.print2 "find trigger for : %s (%s)\n" (Print.term_to_string t) (Print.tag_of_term (SS.compress t));
    
    match (SS.compress t).n with
    | Tm_bvar x ->
      (false, [])

    | Tm_name x ->
      (false, [])

    | Tm_fvar fv ->
      (false, [])

    | Tm_uinst(e, universes) ->
      find_trigger env names e
     
    | Tm_constant c ->
      (false, [])

    | Tm_type u ->
      (true, [])

    | Tm_abs(xs, body, _) -> //fun x1 .. xn -> body
      (true, [])

    | Tm_arrow(xs, body) ->
      (true, [])

    | Tm_refine(x, phi) ->
      (true, [])

    | Tm_app (e, args) ->
      // make sure e doesn't have [@smt_theory_symbol]
      // compute trigger for each args
      // if an arg has a trigger that mention all the bv, then its a trigger
      // if there is no trigger from the args then
      // if none of the args is a trigger killer, then Tm_app is a trigger
      
      //if (FStar.TypeChecker.Env.lookup_attrs_of_lid)
      begin
      match (SS.compress e).n with
      | Tm_fvar fv
      | Tm_uinst({n=Tm_fvar fv}, _)
        when (S.fv_eq_lid fv Const.forall_lid
          ||  S.fv_eq_lid fv Const.exists_lid) ->
        (true, [])

      | _ ->
        let (trigger_killer, _) = find_trigger env names e in
        let (trigger_killer, c) = List.fold_left (fun (k, l) (e, q) -> 
                                      let (b, c) = find_trigger env names e in 
                                      ((k || b), l@c))
                                  (trigger_killer, []) args in
        let c = c |> List.choose (fun t -> filter_trigger names t) in
        let  c = 
          match c with
          | [] -> 
            if (not trigger_killer) && (has_no_smt_theory_symbols env e) then
             // no suitable trigger found yet, add current term as a candidate.
             // Still keep the other candidates around in case we want to infer 
             // patterns with ';'
             c@[t]
            else c
          | _ -> c
        in 
        (trigger_killer, c)
      end
    | Tm_match(e, branches) ->
      (true, [])

    | Tm_ascribed _ -> // what is an ascribed
      (true, [])

    | Tm_let((is_rec, source_lbs), body) ->
      (true, [])

    | Tm_uvar (u, _) -> 
      (false, [])

    | Tm_delayed _ ->
      failwith "Tm_delayed is impossible after compress"

    | Tm_meta(e, m) ->
      (true, [])

    | Tm_lazy i ->
      (true, [])

    | Tm_quoted (tm, qi) ->
      (true, [])

    | Tm_unknown -> 
      (true, [])

(* let check_no_smt_theory_symbols env t =
  match (SS.compress t).n with
  | Tm_fvar fv ->
      if Env.fv_has_attr en fv Const.smt_theory_symbol_attr_lid then [t]
      else []
*)

let terms_to_bvs names =
    match names with
    | [] -> failwith "Empty bound vars"
    | hd::tl ->
        List.fold_left (fun out x -> BU.set_union out (Free.names x)) (Free.names hd) tl

let infer_pattern env (names: list<S.term>) (t:S.term) : list<S.args> =
    let debug = true in
    if debug then 
      BU.print3 "Infer pattern for : %s (%s) with names: (%s)\n" 
        (Print.term_to_string t) 
        (Print.tag_of_term (SS.compress t))
        (names |> List.map Print.term_to_string |> String.concat ", ");
    let bvs = terms_to_bvs names in
    let (_, p) = find_trigger env bvs t in
    let pats = p |> List.choose (fun t -> filter_trigger bvs t) in
    if debug then
      BU.print1 "Found patterns: %s\n" (pats |> List.map Print.term_to_string |> String.concat "; ");
    List.fold_left (fun l t -> l@[[S.as_arg t]]) [] pats

let remove_invalid_pattern (names: list<S.term>) (pats: list<S.args>) : list<S.args> =
    match names with
    | [] -> pats
    | _ ->
      let bvs = terms_to_bvs names in
      let pats = List.fold_left(fun l p -> let p = List.choose (fun (t,_) -> filter_trigger bvs t) p in l@p) [] pats in
      List.fold_left (fun l t -> l@[[S.as_arg t]]) [] pats